File: src/gestorAplicación/interno/Metas.java
package gestorAplicación.interno;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.time.Instant;

public class Metas implements Serializable {
	private static final long serialVersionUID = 5L;
	public static final String nombreD = "Metas";
	public String nombre;
	private double cantidad;
	private Date fecha;
	public int id;
	private static final DateFormat DATE_FORMAT = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault());
	private static ArrayList<Metas> metasTotales = new ArrayList<Metas>();
	private ArrayList<Metas> metasTotalesAux = new ArrayList<Metas>();
	private Usuario dueno;

	// FUNCIONALIDAD
	public static Metas metaProxima;
	public static String plazo;

	// CONSTRUCTORES
	public Metas(String nombre, double cantidad, String fecha, int id) throws ParseException {
		this.setId(id);
		this.nombre = nombre;
		this.cantidad = cantidad;
		this.fecha = DATE_FORMAT.parse(fecha);
		Metas.getMetasTotales().add(this);
		id++;
	}

	public Metas(String nombre, double cantidad, int id) {
		this.setId(id);
		this.nombre = nombre;
		this.cantidad = cantidad;
		Metas.getMetasTotales().add(this);
		id++;
	}

	public Metas(String nombre, String fecha, int id) throws ParseException {
		this.setId(id);
		this.nombre = nombre;
		this.fecha = DATE_FORMAT.parse(fecha);
		Metas.getMetasTotales().add(this);
		id++;
	}

	public Metas(double cantidad, String fecha, int id) throws ParseException {
		this.setId(id);
		this.cantidad = cantidad;
		this.fecha = DATE_FORMAT.parse(fecha);
		Metas.getMetasTotales().add(this);
		id++;
	}

	// Metodos de la funcionalidad asesoramiento de inversion.
	public static Metas revisionMetas(Usuario u) {

		Date proximaFecha = u.getMetasAsociadas().get(0).getFecha();
		Metas proximaMeta = u.getMetasAsociadas().get(0);

		if (u.getMetasAsociadas().size() == 1) {
			if (u.getMetasAsociadas().get(0).getFecha() == null) {
				proximaFecha = Date.from(Instant.now());
				proximaMeta = null;
			}
			return u.getMetasAsociadas().get(0);
		}

		// Debemos comparar cada fecha con todas las otras fechas, por eso hay dos
		// loops.
		for (int i = 0; i < u.getMetasAsociadas().size(); i++) {

			for (int e = 0; e < u.getMetasAsociadas().size(); e++) {
				if (u.getMetasAsociadas().get(e).getFecha() == null
						|| u.getMetasAsociadas().get(i).getFecha() == null) {
					continue;
				} else {
					int r = u.getMetasAsociadas().get(e).getFecha().compareTo(u.getMetasAsociadas().get(i).getFecha());

					if (r < 0) {

						int t = u.getMetasAsociadas().get(e).getFecha().compareTo(proximaFecha);

						if (t < 0) {
							proximaFecha = u.getMetasAsociadas().get(e).getFecha();
							proximaMeta = u.getMetasAsociadas().get(e);
						}

						else {
							continue;
						}
					} else {
						continue;

					}
				}

			}
		}
		return proximaMeta;
	}

	public static Metas cambioFecha(Metas meta, String Fecha) {
		SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
		try {
			Date nuevaFecha = sdf.parse(Fecha);
			meta.setFecha(nuevaFecha);
		} catch (ParseException e) {
			e.printStackTrace();
		}
		return meta;
	}

	public static void determinarPlazo(Metas me1) throws ParseException {
		SimpleDateFormat sdformat = new SimpleDateFormat("yyyy-MM-dd");
		Date d1 = sdformat.parse("2024-01-01");
		Date d2 = sdformat.parse("2026-01-01");
		if (me1.getFecha().compareTo(d1) < 0) {
			plazo = "Corto";
		} else if (me1.getFecha().compareTo(d1) > 0 && me1.getFecha().compareTo(d2) < 0) {
			plazo = "Mediano";
		} else {
			plazo = "Largo";
		}
	}

	public static void prioridadMetas(Usuario u, Metas me) {
		u.getMetasAsociadas().remove(u.getMetasAsociadas().size() - 1);
		u.getMetasAsociadas().add(0, me);
	}

	@Override
	public void finalize() {
		String name = this.getNombre();
		double amount = this.getCantidad();
		Date date = this.getFecha();
		if (date == null) {
			System.out
					.print("La meta con nombre " + this.getNombre() + " fue eliminada satisfactoriamente del sistema.");
		} else if (amount == 0) {
			System.out.print("La meta con nombre " + this.getNombre() + "para la fecha " + this.getFechaNormal()
					+ " fue eliminada satisfactoriamente del sistema.");
		} else if (name == null) {
			System.out.print("La meta para la fecha " + this.getFechaNormal()
					+ " fue eliminada satisfactoriamente del sistema.");
		} else {
			System.out.print("La meta con nombre " + this.getNombre() + "para la fecha " + this.getFechaNormal()
					+ " fue eliminada satisfactoriamente del sistema.");
		}
	}

	// GETTER Y SETTER
	public String getNombre() {
		return nombre;
	}

	public void setNombre(String nombre) {
		this.nombre = nombre;
	}

	public double getCantidad() {
		return cantidad;
	}

	public void setCantidad(double cantidad) {
		this.cantidad = cantidad;
	}

	public String getFechaNormal() {
		SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
		return sdf.format(fecha);
	}

	public Date getFecha() {
		return fecha;
	}

	public void setFecha(Date fecha) {
		this.fecha = fecha;
	}

	public int getId() {
		return id;
	}

	public void setId(int id) {
		this.id = id;
	}

	public static ArrayList<Metas> getMetasTotales() {
		return metasTotales;
	}

	public static void setMetasTotales(ArrayList<Metas> metasTotales) {
		Metas.metasTotales = metasTotales;
	}

	public Usuario getDueno() {
		return dueno;
	}

	public void setDueno(Usuario dueno) {
		this.dueno = dueno;
	}

	public ArrayList<Metas> getMetasTotalesAux() {
		return metasTotalesAux;
	}

	public void setMetasTotalesAux(ArrayList<Metas> metasTotalesAux) {
		this.metasTotalesAux = metasTotalesAux;
	}
}

File: Python/src/gestorAplicación/interno/metas.py
from datetime import date

class Metas:
    _metasTotales = []
    _metaProxima = None
    _plazo = ""
    def __init__(self,**kwargs):
        self._nombre = None
        self.cantidad = None
        self._fecha = None
        self.dueno = None
        Metas._metasTotales.append(self)
        self._id = len(Metas._metasTotales)

        for key in kwargs:
            if key == "nombre":
                self._nombre = kwargs[key]
            if key == "cantidad":
                self.cantidad = kwargs[key]
            if key == "fecha":
                self._fecha = kwargs[key]
            if key == "dueno":
                self.dueno = kwargs[key]

    # Metodos asesoramiento Inversiones
    def revision_metas(user):
        proximaFecha = date.now()
        proximaMeta = None

        if user.getMetasAsociadas().size()==1:
            return user.getMetasAsociadas()[0]
        else:
            for meta in user.getMetasAsociadas():
                if meta.getFecha() != None:
                    proximaMeta = meta
                    break
                else:
                        continue

            for meta1 in user.getMetasAsociadas():
                for meta2 in user.getMetasAsociadas():
                    if meta2.getFecha() is None or meta1.getFecha() is None:
                        continue
                    else:
                        if meta2.getFecha() < meta1.getFecha():
                            if meta2.getFecha() < proximaFecha:
                                proximaFecha = meta2.getFecha()
                                proximaMeta = meta2
                            else:
                                continue
                        else:
                            continue

            return proximaMeta

    @staticmethod
    def cambio_fecha(meta, fecha):
        nuevaFecha = date.date.strptime(fecha, "%d/%m/%Y")
        meta.setFecha(nuevaFecha)
        return meta

    @staticmethod
    def determinar_plazo(meta):
        date1 = date.strptime("01/01/2024", "%d/%m/%Y")
        date2 = date.strptime("01/01/2026", "%d/%m/%Y")

        if meta.getFecha() < date1:
            Metas.plazo = "Corto"
        elif date1 < meta.getFecha() < date2:
            Metas.plazo = "Mediano"
        else:
            Metas.plazo = "Largo"

    @staticmethod
    def prioridad_metas(user, meta):
        user.getMetasAsociadas().pop(len(user.getMetasAsociadas) - 1)
        user.getMetasAsociadas().insert(0, meta)

    @staticmethod
    def limpiar_propiedades(arreglo):
        arreglo.remove("serialVersionUID")
        arreglo.remove("nombreD")
        arreglo.remove("metasTotales")
        arreglo.remove("DATE_FORMAT")
        arreglo.remove("metaProxima")
        arreglo.remove("plazo")
        
# Sets Y Gets
    
    def getNombre(self):
        return self._nombre
    def getCantidad(self):
        return self.cantidad
    def getFecha(self):
        return self._fecha
    def getDueno(self):
        return self.dueno
    def getMetasTotales():
        return Metas._metasTotales
    def getId(self):
        return self._id
    def getMetaProxima():
        return Metas._metaProxima
    def getPlazo():
        return Metas._plazo

    def setNombre(self,nombre):
        self._nombre = nombre 
    def setCantidad(self,cantidad):
        self.cantidad = cantidad
    def setFecha(self,fecha):
        self._fecha = fecha
    def setDueno(self,dueno):
        self.dueno = dueno
    def setMetasTotales(MetasTotales):
        Metas._metasTotales = MetasTotales
    def setId(self,id):
        self._id = id
    def setMetaProxima(metaProxima):
        Metas._metaProxima = metaProxima
    def setPlazo(plazo):
        Metas._plazo = plazo

        